// weapp-tailwindcss@3.7.0/dist/constants-CFM1J-h1.js
'use strict';

const queryKey = 'weapp-tw-platform';
function createMediaQuery(value) {
    return `@media (${queryKey}:"${value}"){&}`;
}
function createNegativeMediaQuery(value) {
    return `@media not screen and (${queryKey}:"${value}"){&}`;
}
function normalComment(text) {
    if (typeof text === 'string') {
        return text.replaceAll(/(?<!\\)_/g, ' ');
    }
    return text;
}
function ifdef(text) {
    return {
        start: `#ifdef ${normalComment(text)}`,
        end: `#endif`,
    };
}
function ifndef(text) {
    return {
        start: `#ifndef ${normalComment(text)}`,
        end: `#endif`,
    };
}
function matchCustomPropertyFromValue(str, cb) {
    let arr;
    let index = 0;
    const regex = new RegExp(`\\(\\s*${queryKey}\\s*:\\s*"([^)]*)"\\)`, 'g');
    while ((arr = regex.exec(str)) !== null) {
        cb(arr, index);
        index++;
    }
}

exports.createMediaQuery = createMediaQuery;
exports.createNegativeMediaQuery = createNegativeMediaQuery;
exports.ifdef = ifdef;
exports.ifndef = ifndef;
exports.matchCustomPropertyFromValue = matchCustomPropertyFromValue;
