// weapp-tailwindcss@3.7.0/dist/defaults-CdqzSWsR.js
'use strict';

require('@ast-core/escape');
var escape = require('@weapp-core/escape');
var defu = require('./defu-Cdz2PomB.js');

function isMap(value) {
    return Object.prototype.toString.call(value) === '[object Map]';
}
function noop() { }
function groupBy(arr, cb) {
    if (!Array.isArray(arr)) {
        throw new TypeError('expected an array for first argument');
    }
    if (typeof cb !== 'function') {
        throw new TypeError('expected a function for second argument');
    }
    const result = {};
    for (const item of arr) {
        const bucketCategory = cb(item);
        const bucket = result[bucketCategory];
        if (Array.isArray(bucket)) {
            result[bucketCategory].push(item);
        }
        else {
            result[bucketCategory] = [item];
        }
    }
    return result;
}
function getGroupedEntries(entries, options) {
    const { cssMatcher, htmlMatcher, jsMatcher, wxsMatcher } = options;
    const groupedEntries = groupBy(entries, ([file]) => {
        if (cssMatcher(file)) {
            return 'css';
        }
        else if (htmlMatcher(file)) {
            return 'html';
        }
        else if (jsMatcher(file) || wxsMatcher(file)) {
            return 'js';
        }
        else {
            return 'other';
        }
    });
    return groupedEntries;
}
function removeExt(file) {
    return file.replace(/\.[^./]+$/, '');
}
const defuOverrideArray = defu.createDefu((obj, key, value) => {
    if (Array.isArray(obj[key]) && Array.isArray(value)) {
        obj[key] = value;
        return true;
    }
});

const defaultOptions = {
    cssMatcher: file => /.+\.(?:wx|ac|jx|tt|q|c|ty)ss$/.test(file),
    htmlMatcher: file => /.+\.(?:(?:wx|ax|jx|ks|tt|q|ty|xhs)ml|swan)$/.test(file),
    jsMatcher: (file) => {
        if (file.includes('node_modules')) {
            return false;
        }
        return /.+\.[cm]?js?$/.test(file);
    },
    mainCssChunkMatcher: (file, appType) => {
        switch (appType) {
            case 'uni-app': {
                return /^common\/main/.test(file);
            }
            case 'uni-app-vite': {
                return file.startsWith('app') || /^common\/main/.test(file);
            }
            case 'mpx': {
                return file.startsWith('app');
            }
            case 'taro': {
                return file.startsWith('app');
            }
            case 'remax': {
                return file.startsWith('app');
            }
            case 'rax': {
                return file.startsWith('bundle');
            }
            case 'native': {
                return file.startsWith('app');
            }
            case 'kbone': {
                return /^(?:common\/)?miniprogram-app/.test(file);
            }
            default: {
                return true;
            }
        }
    },
    wxsMatcher: () => {
        return false;
    },
    cssPreflight: {
        'box-sizing': 'border-box',
        'border-width': '0',
        'border-style': 'solid',
        'border-color': 'currentColor',
    },
    disabled: false,
    customRuleCallback: noop,
    onLoad: noop,
    onStart: noop,
    onEnd: noop,
    onUpdate: noop,
    customAttributes: {},
    customReplaceDictionary: escape.SimpleMappingChars2String,
    appType: undefined,
    arbitraryValues: {
        allowDoubleQuotes: false,
    },
    cssChildCombinatorReplaceValue: ['view'],
    inlineWxs: false,
    injectAdditionalCssVarScope: false,
    jsPreserveClass: (keyword) => {
        if (keyword === '*') {
            return true;
        }
        return false;
    },
    disabledDefaultTemplateHandler: false,
    cssSelectorReplacement: {
        root: 'page',
        universal: ['view', 'text'],
    },
    babelParserOptions: {
        sourceType: 'unambiguous',
    },
    postcssOptions: {},
    cssRemoveHoverPseudoClass: true,
};

exports.defaultOptions = defaultOptions;
exports.defuOverrideArray = defuOverrideArray;
exports.getGroupedEntries = getGroupedEntries;
exports.isMap = isMap;
exports.removeExt = removeExt;
