// weapp-tailwindcss@3.7.0/dist/index-DoMb0-xW.mjs
import { g as getOptions, _ as __awaiter } from './options-yXyCFCtr.mjs';
import { Buffer } from 'node:buffer';
import stream from 'node:stream';
import { c as createDebug } from './index-BEIjgKpF.mjs';

const debug = createDebug();
const Transform = stream.Transform;
function createPlugins(options = {}) {
    const opts = getOptions(options);
    const { templateHandler, styleHandler, patch, jsHandler, setMangleRuntimeSet, cache, twPatcher } = opts;
    let runtimeSet = new Set();
    patch === null || patch === void 0 ? void 0 : patch();
    function transformWxss(options = {}) {
        return new Transform({
            objectMode: true,
            transform(file, _encoding, callback) {
                return __awaiter(this, void 0, void 0, function* () {
                    runtimeSet = twPatcher.getClassSet();
                    setMangleRuntimeSet(runtimeSet);
                    const error = null;
                    if (file.contents) {
                        const rawSource = file.contents.toString();
                        const hash = cache.computeHash(rawSource);
                        cache.calcHashValueChanged(file.path, hash);
                        yield cache.process(file.path, () => {
                            const source = cache.get(file.path);
                            if (source) {
                                file.contents = Buffer.from(source);
                                debug('css cache hit: %s', file.path);
                            }
                            else {
                                return false;
                            }
                        }, () => __awaiter(this, void 0, void 0, function* () {
                            const { css } = yield styleHandler(rawSource, Object.assign({ isMainChunk: true }, options));
                            file.contents = Buffer.from(css);
                            debug('css handle: %s', file.path);
                            return {
                                key: file.path,
                                source: css,
                            };
                        }));
                    }
                    callback(error, file);
                });
            },
        });
    }
    function transformJs(options = {}) {
        return new Transform({
            objectMode: true,
            transform(file, _encoding, callback) {
                return __awaiter(this, void 0, void 0, function* () {
                    const error = null;
                    if (file.contents) {
                        const rawSource = file.contents.toString();
                        const hash = cache.computeHash(rawSource);
                        cache.calcHashValueChanged(file.path, hash);
                        yield cache.process(file.path, () => {
                            const source = cache.get(file.path);
                            if (source) {
                                file.contents = Buffer.from(source);
                                debug('js cache hit: %s', file.path);
                            }
                            else {
                                return false;
                            }
                        }, () => __awaiter(this, void 0, void 0, function* () {
                            const { code } = yield jsHandler(rawSource, runtimeSet, options);
                            file.contents = Buffer.from(code);
                            debug('js handle: %s', file.path);
                            return {
                                key: file.path,
                                source: code,
                            };
                        }));
                    }
                    callback(error, file);
                });
            },
        });
    }
    function transformWxml(options = {}) {
        return new Transform({
            objectMode: true,
            transform(file, _encoding, callback) {
                return __awaiter(this, void 0, void 0, function* () {
                    const error = null;
                    if (file.contents) {
                        const rawSource = file.contents.toString();
                        const hash = cache.computeHash(rawSource);
                        cache.calcHashValueChanged(file.path, hash);
                        yield cache.process(file.path, () => {
                            const source = cache.get(file.path);
                            if (source) {
                                file.contents = Buffer.from(source);
                                debug('html cache hit: %s', file.path);
                            }
                            else {
                                return false;
                            }
                        }, () => __awaiter(this, void 0, void 0, function* () {
                            const code = yield templateHandler(rawSource, Object.assign({ runtimeSet }, options));
                            file.contents = Buffer.from(code);
                            debug('html handle: %s', file.path);
                            return {
                                key: file.path,
                                source: code,
                            };
                        }));
                    }
                    callback(error, file);
                });
            },
        });
    }
    return {
        transformWxss,
        transformWxml,
        transformJs,
    };
}

export { createPlugins as c };
